### PR TITLE
Iterative merged update

### PR DESCRIPTION
**Pull request by roironn**
_Chaned workspace and multi env sources to avoid reading the entire sub sources and use the changes returned by the sub sources for init instead_

---

---
_Release Notes_: 
_NA_

### CODE REVIEW

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 1]
```diff
@@ -92,59 +96,38 @@ const buildMultiEnvSource = (
...
+    changes?: EnvChanges
   }): Promise<{ state: MultiEnvState; changes: Change<Element>[] }> => {
     const primaryEnv = env ?? primarySourceName
-    if (primaryEnv !== primarySourceName) {
-      return { state: await buildState(env), changes: [] }
-    }
-    const current = state !== undefined ? await state : await buildState(env)
+    const current = states[primaryEnv] !== undefined
+      ? states[primaryEnv]
+      : await buildState(env)
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:130]: Shouldn't this be `await buildState(primaryEnv)`?

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 2]
```diff
@@ -154,30 +137,88 @@ const buildMultiEnvSource = (
       id => id.getFullName()
     )
 
-    const changedElementsByEnv = _.mapValues(
+
+    const envChangesByID = _.mapValues(
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:141]: Naming is misleading here cause it's byEnvById and not really only by Id

Also, the values here are a mapping between ElemIDFullname to an Element. An inMemoryElemntsSource is an overkil?

Not your change but 'envs' is also mis-leading a bit. It's maybe envSources (cause it's primary+common)

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 3]
```diff
@@ -154,30 +137,88 @@ const buildMultiEnvSource = (
...
+      relevantElementIDs
+        .map(id => envChangesByID[environment][id.getFullName()])
+        .filter(values.isDefined)
+        .map(getAfterFromChange)
+        .filter(values.isDefined)
+    )).flat()
+
+    const unmodifiedFragments = awu(relevantElementIDs).map(async id => {
+      const primChange = envChangesByID[primaryEnv][id.getFullName()]
+      const commonChange = envChangesByID[primaryEnv][id.getFullName()]
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:157]: Shouldn't this be `const commonChange = envChangesByID[commonSourceName][id.getFullName()]`?

File: packages/workspace/src/workspace/remote_map.ts [hunk 4]
```diff
@@ -115,3 +115,17 @@ export class InMemoryRemoteMap<T, K extends string = string> implements RemoteMa
 
   [Symbol.toStringTag]: '[InMemoryRemoteMap]'
 }
+
+export const mapRemoteMapResult = <T>(
```
Comment (adamro) [inline packages/workspace/src/workspace/remote_map.ts:119]: Name was a bit confusing but I don't have a much better name

I guess you didn't have a use-case but I feel this could also be useful with T and K? ie. `func: (orig: T) => Promise<T>` and `RemoteMap<K>` return value

File: packages/workspace/test/workspace/elements_source.test.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ import {
   createInMemoryElementSource, RemoteElementSource,
 } from '../../src/workspace/elements_source'
 
+
```
Comment (adamro) [inline packages/workspace/test/workspace/elements_source.test.ts:21]: remove :)

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 6]
```diff
@@ -438,40 +438,59 @@ const buildNaclFilesSource = (
...
+        parsedNaclFiles[filename] = await toParsedNaclFile(naclFile, parseResult)
       }
     })
-
-    const cacheOnlyState = (await buildNaclFilesState({
-      newNaclFiles: parsedNaclFilesFromCache,
-      remoteMapCreator,
-      staticFilesSource,
-      sourceName,
+    await awu(fileNames).forEach(async filename => {
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:454]:
```
const modifiedNaclFiles = (await (awu(fileNames).map(async filename => {
      const naclFile = (await naclFilesStore.get(filename) ?? { filename, buffer: '' }) as NaclFile
      return await cache.hasValid(cacheResultKey(naclFile)) ? undefined : naclFile
    })).toArray()).filter(values.isDefined)
```

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 7]
```diff
@@ -438,40 +438,59 @@ const buildNaclFilesSource = (
...
-        const latestCache = validCache ?? await cache.get(cacheResultKey(naclFile), true)
-        if (latestCache !== undefined) {
-          parsedNaclFilesFromCache.push(await toParsedNaclFile(naclFile, latestCache))
-        }
-        if (validCache === undefined) {
-          modifiedNaclFiles.push(naclFile)
-        }
+    const fileNames = new Set([...cacheFilenames, ...naclFilenames])
+    const parsedNaclFiles: Record<string, ParsedNaclFile> = {}
+    await awu(cacheFilenames).forEach(async filename => {
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:447]: Why did you change this from one iteration to 2?

I have in my branch

```
    const fileNames = new Set([...cacheFilenames, ...naclFilenames])
    await awu(fileNames).forEach(async filename => {
      const naclFile = await naclFilesStore.get(filename) ?? { filename, buffer: '' }
      const cacheVal = await cache.get(filename)
      if (cacheVal !== undefined) {
        parsedNaclFilesFromCache.push(cacheVal)
      }
      const cacheHasValidVal = await cache.hasValid(cacheResultKey(naclFile))
      if (!cacheHasValidVal) {
        modifiedNaclFiles.push(naclFile)
      }
    })
```

which does both in one (and if it's two iterations, why the push :D )

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 8]
```diff
@@ -58,47 +56,49 @@ export type NaclFile = {
   timestamp?: number
 }
 
-export type NaclFilesSource = Omit<ElementsSource, 'clear'> & {
-  updateNaclFiles: (changes: DetailedChange[], mode?: RoutingMode) => Promise<Change<Element>[]>
+export type NaclFilesSource<CHANGES=Change<Element>[]> = Omit<ElementsSource, 'clear'> & {
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:59]: CAPS

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts (hunk 1 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:143]: Fixed it - but all of the method that uses down the line fallback to the primary env so this was actually ok...

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts (hunk 2 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:154]: changed to envsChangesByID

File: packages/adapter-api/src/change.ts [hunk 9]
```diff
@@ -64,6 +64,11 @@ export const isRemovalOrModificationChange = <T extends Change<unknown>>(
 export const getChangeElement = <T>(change: Change<T>): T =>
   (change.action === 'remove' ? change.data.before : change.data.after)
 
+export const getAfterFromChange = <T>(change: Change<T>): T | undefined => {
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/adapter-api/src/change.ts:67]: It just checks that the change is a modification or an addition... TS won't let you access .after without it cause of the way we defined the types...

File: packages/cli/src/command_builder.ts [hunk 10]
```diff
@@ -221,7 +221,6 @@ export const createWorkspaceCommand = <T>(
       args.workspacePath,
       getConfigOverrideChanges(args.input),
     )
-
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/cli/src/command_builder.ts:224]: Point taken.

File: packages/lang-server/src/workspace.ts [hunk 11]
```diff
@@ -224,12 +224,13 @@ export class EditorWorkspace {
...
       // We start by running all deleted
       const removeChanges = (!_.isEmpty(opDeletes))
-        ? await this.workspace.removeNaclFiles(...opDeletes)
+        ? (await this.workspace.removeNaclFiles(...opDeletes))
         : []
       // Now add the waiting changes
       const updateChanges = (!_.isEmpty(opUpdates))
-        ? await this.workspace.setNaclFiles(...Object.values(opUpdates))
-        : []
+        ? (
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/lang-server/src/workspace.ts:231]: Left over from debug - will remove

File: packages/adapter-api/src/change.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (adamro) [reply packages/adapter-api/src/change.ts:67]: We have a `isAdditionOrModificationChange` for that

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 12]
```diff
@@ -74,3 +75,54 @@ export const buildNewMergedElementsAndErrors = async ({
   })
   return changes
 }
+
+export const getBuildMergeData = async ({
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:79]: This prepares the "after" part for creating merged changes. So it's more a `mergeAfterElements` or `buildElementsAfterMapping`.

Also, the change type can be a map between elemID (fullName) to the afterElement. With undefined if it's a remove one. So it will be more straight forwards what this is.

File: packages/workspace/src/workspace/hidden_values.ts [hunk 13]
```diff
@@ -383,36 +387,101 @@ const isHiddenField = async (
       || isHiddenField(baseType, fieldPath.slice(0, -1), hiddenValue, elementsSource)
 }
 
+const diffElements = <T extends Element>(fullElem: T, visibleElem: T): T | undefined => {
+  if (fullElem === undefined) {
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:391]: nit: undefined doesn't extend Element so the two input changes shouldn't really be possible. I am guessing this was a test/bug fix cause one of the usages has casting or didn't filter properly or w/e. If it's required I would change to `T | undefined` or add a ? in the func params I guess.

File: packages/workspace/src/workspace/hidden_values.ts [hunk 14]
```diff
@@ -383,36 +387,101 @@ const isHiddenField = async (
       || isHiddenField(baseType, fieldPath.slice(0, -1), hiddenValue, elementsSource)
 }
 
+const diffElements = <T extends Element>(fullElem: T, visibleElem: T): T | undefined => {
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:390]: Just to align it with `addedDiff` I would change the order 

`(visibleElement: T, fullElement: T)`

File: packages/workspace/src/workspace/hidden_values.ts [hunk 15]
```diff
@@ -383,36 +387,101 @@ const isHiddenField = async (
...
+  if (fullElem === undefined) {
+    return undefined
+  }
+  if (visibleElem === undefined) {
+    return fullElem
+  }
+  const diffAnno = addedDiff(visibleElem.annotations, fullElem.annotations)
+  const diffAnnoTypes = {}
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:398]: Better to create it typed as ReferenceMap from start instead than casting in the end imo

File: packages/workspace/src/workspace/hidden_values.ts [hunk 16]
```diff
@@ -383,36 +387,101 @@ const isHiddenField = async (
...
+  if (isObjectType(fullElem) && isObjectType(visibleElem)) {
+    const diffFields = _.pickBy(_.mapValues(
+      fullElem.fields,
+      (field, name) => diffElements(field, visibleElem.fields[name])
+    ), values.isDefined)
+    return [diffAnno, diffAnnoTypes, diffFields].every(_.isEmpty) ? undefined : new ObjectType({
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:404]: isn't `diffAnnoTypes` definitely empty at this stage?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 17]
```diff
@@ -383,36 +387,101 @@ const isHiddenField = async (
...
+      annotations: diffAnno,
+      path: fullElem.path,
+    }) as unknown as T
+  }
+  if (isInstanceElement(fullElem) && isInstanceElement(visibleElem)) {
+    const diffValue = addedDiff(visibleElem.value, fullElem.value)
+    if ([diffAnno, diffAnnoTypes, diffValue].every(_.isEmpty)) {
+      return undefined
+    }
+    const res = fullElem.clone() as InstanceElement
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:427]: This cast is needed?

File: packages/adapter-utils/src/utils.ts [hunk 18]
```diff
@@ -832,20 +832,25 @@ const createDefaultValuesFromType = async (
     : type.annotations[CORE_ANNOTATIONS.DEFAULT])
 }
 
+export const applyInstanceDefaults = async (
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/adapter-utils/src/utils.ts:835]: It actually like this for improved efficiency...  without this you need to separate elements, apply defaults to instances, then concat. 

This way its just a map... So less things to do. Rename maybe?

[the remark this answers is not in the document]
Comment (roironn) [reply packages/adapter-utils/src/utils.ts:835]: (Also - in order to split the elements array you need a to evaluate the entire iterable which is less desirable...

[the remark this answers is not in the document]
Comment (adamro) [reply packages/adapter-utils/src/utils.ts:835]: Even if you want to do it in a single map iteration without the filter complexity you can do the isInstanceElement outside of this function

Btw, we might want a splitBy func in awu ;)

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 19]
```diff
@@ -326,13 +342,20 @@ const buildMultiEnvSource = (
...
-  const load = async (ignoreFileChanges = false): Promise<Change[]> => {
-    const changes = await mapValuesAsync(getActiveSources(), src => src.load(ignoreFileChanges))
-    const buildRes = await buildMultiEnvState({ changes })
-    state = Promise.resolve(buildRes.state)
-    // We still need to load the secondrary nacl file sources
-    await awu(Object.values(secondarySources())).forEach(src => src.load(ignoreFileChanges))
-    return buildRes.changes
+  const load = async (
+    ignoreFileChanges = false,
+    _cachePrefix?: string,
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:347]: remove if not needed?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 20]
```diff
@@ -479,7 +481,7 @@ const buildNaclFilesSource = (
     }
     const naclFile = await getNaclFile(filename)
     if (naclFile === undefined) return undefined
-    return (await parseNaclFiles([naclFile], cache, functions))[0]
+    return (await parseNaclFiles([naclFile], (await getState()).parsedNaclFiles, functions))[0]
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:484]: Not you (actually my fault :) but if we only get here when there's no state and thus no cache, you can use `parseNaclFile` and `toParsedNaclFile` here instead of `parseNaclFiles` with the array trick and the extra cache calls

File: packages/workspace/src/workspace/hidden_values.ts (hunk 17 above)
Comment (roironn) [reply packages/workspace/src/workspace/hidden_values.ts:427]: Its an issue with the clone... Removed it in favour of creating a new element

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 21]
```diff
@@ -106,92 +110,102 @@ const buildMultiEnvSource = (
...
       currentErrors: current.mergeErrors,
       relevantElementIDs: awu(relevantElementIDs),
-      mergeFunc: elements => mergeElements(elements),
+      mergeFunc: async elements => {
+        const plainResult = await mergeElements(elements)
+        return {
+          errors: plainResult.errors,
+          merged: mapRemoteMapResult(plainResult.merged,
+            async element => {
+              const r = await applyInstanceDefaults(
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:162]: noice

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 22]
```diff
@@ -717,29 +719,56 @@ const buildNaclFilesSource = (
...
-        await cache.clear()
+        const currentState = await getState()
+        await currentState.elementsIndex.clear()
+        await currentState.mergeErrors.clear()
+        await currentState.mergedElements.clear()
+        await currentState.referencedIndex.clear()
+        await currentState.parsedNaclFiles.clear()
       }
-      state = undefined
+      // state = Promise.resolve(await buildInitState(cachePrefix)).state
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:729]: The state = undefined was to "trigger" it to fill it again? It will happen now without it?

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts (hunk 19 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:347]: I'll change the signature to accept params object instead and have this optional

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 23]
```diff
@@ -717,29 +719,56 @@ const buildNaclFilesSource = (
...
     },
 
     rename: async (name: string) => {
       await naclFilesStore.rename(name)
       await staticFilesSource.rename(name)
-      await cache.rename(name)
+      const currentState = await getState()
+      await currentState.parsedNaclFiles.rename(name)
+      const newCurrentState = await createNaclFilesState(
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:737]: Maybe allow the `createNaclFilesState` to be able to accept an optional `ParsedNaclFilesCache` and then you can avoid the next line

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 24]
```diff
@@ -717,29 +719,56 @@ const buildNaclFilesSource = (
...
+      )
+      await currentState.elementsIndex.clear()
+      await newCurrentState.mergeErrors.setAll(
+        currentState.mergeErrors.entries()
+      )
+      await currentState.mergeErrors.clear()
+      await newCurrentState.referencedIndex.setAll(
+        currentState.referencedIndex.entries()
+      )
+      await currentState.referencedIndex.clear()
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:756]: I am a bit confused with how this actually changes the next call to `getState()`, shouldn't this have a `state = newCurrentState` line in the end? I probably missed something

File: packages/workspace/src/workspace/workspace.ts [hunk 25]
```diff
@@ -545,9 +600,15 @@ export const loadWorkspace = async (
       if (persist) {
         await config.setWorkspaceConfig(workspaceConfig)
       }
-      naclFilesSource = multiEnvSource(_.mapValues(elementsSources.sources, e => e.naclFiles),
-        currentEnv(), elementsSources.commonSourceName, remoteMapCreator)
-      workspaceState = undefined
+      // naclFilesSource.setCurrentEnv(env)
```
Comment (adamro) [inline packages/workspace/src/workspace/workspace.ts:603]: r / comment

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts (hunk 19 above)
Comment (adamro) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:347]: A more general q, in what case is it optional?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 22 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:729]: It was to clear the parsed nacl files so that they will be recalculated - we now clear them explicitly.

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 23 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:737]: 👍

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 24 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:756]: 👍 
Was a bug - tests only checked the the clear calls was invoked, I'll add a test to make sure the state has changed.

File: packages/workspace/src/workspace/workspace.ts (hunk 25 above)
Comment (roironn) [reply packages/workspace/src/workspace/workspace.ts:603]: 👍

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 26]
```diff
@@ -73,7 +73,9 @@ type MultiEnvState = {
   mergeErrors: RemoteMap<MergeError[]>
 }
 
-type MultiEnvSource = Omit<NaclFilesSource, 'getAll' | 'getElementsSource'> & {
+export type EnvChanges = Record<string, Change[]>
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:76]: I think that defining a type as a string alias is a bit of an abuse DYT?
With change the type name though

[the remark this answers is not in the document]
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:76]: 👍  lol

[the remark this answers is not in the document]
Comment (adamro) [reply packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:76]: Don't think it's an abuse. It's 100% to improve readabilty but if the name is good enough it's not that important.

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 20 above)
Comment (roironn) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:484]: I still prefer to use the regular method - since if in the future we will make changes for the way we use the cache in correlation with this we won't miss this spot... WdYT?

Comment (adamro) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:489]: Looking at this again, how is it even working?

`await getState()` whilst `state === undefined` (and we only get here in this case) should throw an Error

File: packages/workspace/src/workspace/hidden_values.ts [hunk 27]
```diff
@@ -384,23 +383,23 @@ const isHiddenField = async (
...
+const diffElements = <T extends Element>(visibleElem?: T, fullElem?: T): T | undefined => {
   if (fullElem === undefined) {
     return undefined
   }
   if (visibleElem === undefined) {
     return fullElem
   }
   const diffAnno = addedDiff(visibleElem.annotations, fullElem.annotations)
-  const diffAnnoTypes = {}
+  const diffAnnoTypes: ReferenceMap = {}
```
Comment (adamro) [inline packages/workspace/src/workspace/hidden_values.ts:394]: Add a comment that annoTypes can not be hidden so it should always be {}

**Review by adamro — APPROVED** [review]
